util/winutil/restartmgr_windows.go

```diff
@@ -0,0 +1,830 @@
+	AddPaths(fqPaths []string) error
+	// AffectedProcesses returns the UniqueProcess information for all running
+	// processes that utilize the binaries previously specified by calls to
+	// AddPaths.
+	AffectedProcesses() ([]UniqueProcess, error)
+	// Key returns the session key associated with this instance.
+	Key() string
+}
+
+// rmSession encapsulates the necessary information to represent an open
+// restart manager session.
+//
+// Implementation note: rmSession methods that return errors should use
+// windows.Errno codes whenever possible, as we call them from the custom
+// action DLL. MSI custom actions are expected to return windows.Errno values;
+// to ensure our compliance with this expectation, we should also use those
+// values. Failure to do so will result in a generic windows.Errno being
+// returned to the Windows Installer, which obviously is less than ideal.
+type rmSession struct {
+	session _RMHANDLE
+	key     string
+	logf    logger.Logf
+}
+
+const _CCH_RM_SESSION_KEY = 32 // (excludes NUL terminator)
+
+// NewRestartManagerSession creates a new RestartManagerSession that utilizes
+// logf for logging.
+func NewRestartManagerSession(logf logger.Logf) (RestartManagerSession, error) {
+	var sessionKeyBuf [_CCH_RM_SESSION_KEY + 1]uint16
+	result := rmSession{
+		logf: logf,
+	}
+	if err := rmStartSession(&result.session, 0, &sessionKeyBuf[0]); err != nil {
+		return nil, err
+	}
+
+	result.key = windows.UTF16ToString(sessionKeyBuf[:len(sessionKeyBuf)-1])
+	return &result, nil
+}
+
+// AttachRestartManagerSession opens a connection to an existing session
+// specified by sessionKey, using logf for logging.
+func AttachRestartManagerSession(logf logger.Logf, sessionKey string) (RestartManagerSession, error) {
+	sessionKey16, err := windows.UTF16PtrFromString(sessionKey)
+	if err != nil {
+		return nil, err
+	}
+
+	result := rmSession{
+		key:  sessionKey,
+		logf: logf,
+	}
+	if err := rmJoinSession(&result.session, sessionKey16); err != nil {
+		return nil, err
+	}
+	return &result, nil
+}
+
+func (rms *rmSession) Close() error {
+	// 0 is a valid value for rms.session, so we cannot do much here other than
+	// just calling rmEndSession and returning its error.
+	return rmEndSession(rms.session)
+}
+
+func (rms *rmSession) Key() string {
+	return rms.key
+}
+
+func (rms *rmSession) AffectedProcesses() ([]UniqueProcess, error) {
+	infos, err := rms.processList()
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]UniqueProcess, 0, len(infos))
+	for _, info := range infos {
+		result = append(result, UniqueProcess{
+			_RM_UNIQUE_PROCESS: info.Process,
+			CanReceiveGUIMsgs:  info.AppType == _RmMainWindow || info.AppType == _RmOtherWindow,
+		})
+	}
+
+	return result, nil
+}
+
+func (rms *rmSession) processList() ([]_RM_PROCESS_INFO, error) {
+	var avail, rebootReasons uint32
+	needed := uint32(1)
+
+	var buf []_RM_PROCESS_INFO
+	err := error(windows.ERROR_MORE_DATA)
+	for err == windows.ERROR_MORE_DATA {
```

Comment:
Microsoft's example code puts a bound on how many times this loop can run, that might be worth considering.